proto/src/serializers/log.rs

```diff
@@ -0,0 +1,35 @@
+//! Serialize and deserialize serde_json::Value
+use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+use std::str::FromStr;
+
+/// Deserialize string into serde_json::Value
+pub fn deserialize<'de, D>(deserializer: D) -> Result<serde_json::Value, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    // Deserialize into String first.
+    // This has the added side-effect that it consumes escaped quotes in sub-structs,
+    // deserializing them properly as Value.
+    let incoming_string = String::deserialize(deserializer)?;
+    if incoming_string.is_empty() {
+        return Ok(serde_json::Value::Null);
+    }
+    // Try to deserialize as-is (usually sequence/array) or as string with extra quotes.
+    serde_json::Value::from_str(&incoming_string).or_else(|e| {
+        serde_json::Value::from_str(&format!("\"{}\"", incoming_string))
+            .map_err(|_| D::Error::custom(format!("{}", e)))
+    })
+}
+
+/// Serialize from serde_json::Value into string
+pub fn serialize<S>(value: &serde_json::Value, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    value.to_string().serialize(serializer)
+}
+
+/// Helper struct to serialize/deserialize log messages
+#[derive(Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct Log(#[serde(with = "crate::serializers::log")] pub serde_json::Value);
```

Comment:
The `log` field is actually just a string. See [here](https://github.com/tendermint/tendermint/blob/master/abci/types/types.pb.go#L1923) and [here](https://github.com/tendermint/tendermint/blob/master/proto/tendermint/abci/types.proto#L204).

Given that we're not concerned with what's inside the log field at the Tendermint-rs level, I'd recommend against modelling it any other way. Rather let consumers interpret the string in an application-dependent way.

Reply:
It was already understood as a JSON struct in Tendermint-rs, albeit it was optional: https://github.com/informalsystems/tendermint-rs/blob/0ad5f50b7368b8eff3b6ff8db323c730704a94b2/tendermint/src/abci/log.rs#L9-L14

If we only parse it as a String, we're going to give hard times for Anca and other downstream projects trying to parse it.

I can reintroduce the `parse_json()` function with some updated code, so they can have it both ways. But my gut tells me if this field is not locked down to a specification, it's going to be abused (which it already is, somewhat)

Reply:
Actually, my main concern came from the struct Anca sent yesterday: it had a weird amount of 2 backslashes for every quote and it took me half a night to realize that it's not the code's fault that it can't deserialize it (even as a string), but it's the string itself that's "overescaped".
From there, I wanted to make sure that a string can be deserialized as a `serde_json::Value`.

Reply:
Yeah that doesn't seem right. All we should be concerned about from the Tendermint-rs perspective is the string representation.

What if a different ABCI app uses a different format in its `log` field? Then they're forced to interpret it as a `serde_json::Value`?

Reply:
> Actually, my main concern came from the struct Anca sent yesterday: it had a weird amount of 2 backslashes for every quote and it took me half a night to realize that it's not the code's fault that it can't deserialize it (even as a string), but it's the string itself that's "overescaped".

Yeah I saw this too, but when I copied/pasted her raw JSON into a Python interpreter I didn't get the weird double-backslashes. What'd you use to format the JSON? From the Python 3 interpreter, I just did:

```python3
s = """... JSON here ..."""
print(s)
```

Python took care of all the newlines, and didn't produce any double backslashes.

Reply:
I can try your fix in ibc with gaia if you think this should work. Just to make sure :) Let me know.

Reply:
Ok, so, Thane, do you want to completely get rid of the `serde_json::Value` parsing? Our options are to keep it (might restrict downstream projects too much), make it optional (keeps the current feature set, but it's an extra feature for unknown benefits), remove it (makes life simple here and moves the problem to downstream projects).

BTW, the one restriction we might keep is that we deserialize it as a `String` == it has to be UTF-8 compatible. I think we can get away with that, it's similar to the Go definition.

Reply:
Again, the log field's just a string, so I personally think there shouldn't be any notion of JSON anywhere near it 😁 (at least in the tendermint-rs codebase)

Reply:
done